### PR TITLE
build: compilation refactor

### PR DESCRIPTION
**Pull request by outscale-avu**

### CODE REVIEW

**Comment by outscale-mgo** [comment]
Can one of the admins verify this patch?

File: benchmark.mk [hunk 1]
```diff
@@ -0,0 +1,200 @@
+BENCHMARK_INCLUDED=true
+
+.PHONY: $(bench_antispoof_OBJECTS) $(bench_core_OBJECTS) $(bench_diode_OBJECTS) $(bench_rxtx_OBJECTS) $(bench_pmtud_OBJECTS) $(bench_ip-fragment_OBJECTS) $(bench_firewall_OBJECTS) $(bench_nic_OBJECTS) $(bench_print_OBJECTS) $(bench_queue_OBJECTS) $(bench_switch_OBJECTS) $(bench_vtep_OBJECTS) $(bench_tap_OBJECTS) $(bench_thread_OBJECTS) $(bench_integration_OBJECTS) $(bench_vhost_OBJECTS)
```
Comment (outscale-mgo) [inline benchmark.mk]: but PHONY should have only rules that are not file right ?
so why are OBJECT rules here ?

File: config.mk [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#####################################
```
Comment (outscale-mgo) [inline config.mk]: this file is auto generated, so it should not be in this commit, but in the .gitignore

File: configure [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	echo -e  "#include <stdio.h>\nint main() {\nprintf(\"hello world\\\n\");\n}" > hw.c
+	$1 -Wall hw.c -o cccheck.out 2> /dev/null
+	retcomp=$?
+	rm hw.c 2> /dev/null
+	retstr=$(./cccheck.out 2> /dev/null)
+	./cccheck.out > /dev/null 2>/dev/null
+	retexec=$?
+	rm ./cccheck.out 2> /dev/null
+	if [ $retcomp -ne 0 ] || [ "$retstr" != "hello world" ] || [ $retexec -ne 0 ]; then
+		echo ">>> Compiler failed"
```
Comment (outscale-mgo) [inline configure]: can you write on stderr ?
also ">>> Compiler failed" seems a little light and why ">>>".
maybe you should add the compiler use
"$CC fail simple compilation test"
you can even use the glorious error_print() (which doesn't output to stderr either, but this is a bug inside this function)

File: configure [hunk 4]
```diff
@@ -0,0 +1,212 @@
+#!/bin/bash
+
+function checkCompiler {
```
Comment (outscale-mgo) [inline configure]: can you use underscore case instead of camlCase please 
we use Linux style which use underscore in C, so why change it for shell script ?

File: configure [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+	./cccheck.out > /dev/null 2>/dev/null
+	retexec=$?
+	rm ./cccheck.out 2> /dev/null
+	if [ $retcomp -ne 0 ] || [ "$retstr" != "hello world" ] || [ $retexec -ne 0 ]; then
+		echo ">>> Compiler failed"
+		exit
+	fi
+}
+
+function printHelp {
```
Comment (outscale-mgo) [inline configure]: print_help

File: configure [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+		echo -e "  -p --print\t\t\tPrint variables values"
+		echo -e "  -P --modifiers\t\tPrint available variables"
+		echo -e "  -b --with-benchmarks\t\tCompile benchmarks"
+		echo -e "  -e --with-examples\t\tCompile examples"
+		echo -e "  -a --asan\t\t\tCompile examples"
+		echo
+		printModifiers
+}
+
+function printModifiers {
```
Comment (outscale-mgo) [inline configure]: print_modifier

File: configure [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+		echo -e "Syntax    : ./configure [options] CC=gcc EXTRA_CFLAGS=\"-g -O2 -j4 ...\""
+		echo -e "  CC\t\tR\tCompiler"
+		echo -e "  EXTRA_CFLAGS\tA\tCompiler option, adds to PG_CFLAGS normal compilation"
+		echo -e "  EXTRA_dev_CFLAGS\tA\tCompiler option, adds to PG_dev_CFLAGS dev options compilation"
+		echo -e "  RTE_SDK\tR\tRoot directory of DPDK"
+		echo -e "  PG_NAME\tR\tName of the compiled library ({NAME}.a / {NAME}.so.x.y.z)"
+		echo -e "  PG_dev_NAME\tR\tName of the compiled library ({NAME}-dev.a / {NAME}-dev.so.x.y.z)"
+}
+
+function printVar {
```
Comment (outscale-mgo) [inline configure]: print_var
and so on, I won't do every functions, but you got the idea...

File: npf.mk [hunk 8]
```diff
@@ -0,0 +1,55 @@
+include npfvar.mk
+
+prop : $(prop_OBJECTS)
+	ar rcv libprop.a $(prop_OBJECTS)
+
```
Comment (outscale-gle) [inline npf.mk]: Care you are in an old version of Packetgraph, think about rebase ! For example Prop is no longer in this world

File: configure [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+addExample=false
+addAsan=false
+newExport=""
+
+echo "#####################################" > config.mk
+echo "### Configuration for packetgraph ###" >> config.mk
+echo "#####################################" >> config.mk
+
+for var in "$@"; do
+	if [[ ${var:0:2} == "--" ]]; then
```
Comment (outscale-mgo) [inline configure]: I guess you can merge checkOption and checkOptionSingle into a single function

File: error_print.sh [hunk 10]
```diff
@@ -9,13 +9,13 @@ error_print()
     has_ponysay=$?
 
     if [ $has_ponysay -eq 0 ]; then
-	ponysay --pony derpysad $1
```
Comment (outscale-mgo) [inline error_print.sh]: re-indentation seems unrelated to this commit

File: configure [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+		echo "Usage   : ./configure [options] [variable]"
+		echo "Options :"
+		echo -e "  -h --help\t\t\tPrint this help and exit"
+		echo -e "  -v --verbose\t\t\tAugment compilation verbose"
+		echo -e "  -n --no-verbose\t\tReduce compilation verbose"
+		echo -e "  -p --print\t\t\tPrint variables values"
+		echo -e "  -P --modifiers\t\tPrint available variables"
+		echo -e "  -b --with-benchmarks\t\tCompile benchmarks"
+		echo -e "  -e --with-examples\t\tCompile examples"
+		echo -e "  -a --asan\t\t\tCompile examples"
```
Comment (outscale-gle) [inline configure]: Care -a don't compile examples but compile with asan.

File: configure-helper.sh [hunk 12]
```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
```
Comment (outscale-mgo) [inline configure-helper.sh]: maybe you can create a commit only for this, rule are:
- try to have one commit per change 
- commit should not be dependent of a "future commit" 
So sure, this commit would add dead code without your build system, but it won't break.
I prefer sweet small commit over big fat diff.
fat shaming... sorry, I'm not sorry I don't like fat commit, they're ugly, and generally unnecessary

Comment (outscale-gle) [reply configure-helper.sh]: You can do this for examples for .mk files. 🗡
